### PR TITLE
Adds Booster Energy

### LINKED ISSUE #4065: Booster Energy Implementation missing
### Description

Booster Energy is missing in functionality.

### Discord contact info

._.terra._.

### PR DESCRIPTION
**Pull request by AlexOn1ine**
Closes #4065

### CODE REVIEW

File: data/battle_scripts_1.s [hunk 1]
```diff
@@ -9926,6 +9926,16 @@ BattleScript_BerserkGeneRet_End:
 	removeitem BS_SCRIPTING
 	end3
 
+BattleScript_BoosterEnergyEnd2::
```
Comment (Bassoonian) [inline data/battle_scripts_1.s:9929]: Is there a reason this is not just `BattleScript_BoosterEnergyEnd`?

File: src/battle_message.c [hunk 2]
```diff
@@ -791,6 +791,7 @@ static const u8 sText_AttackerSwitchedStatWithTarget[] = _("{B_ATK_NAME_WITH_PRE
 static const u8 sText_BeingHitChargedPkmnWithPower[] = _("Being hit by {B_CURRENT_MOVE}\ncharged {B_DEF_NAME_WITH_PREFIX} with power!");
 static const u8 sText_SunlightActivatedAbility[] = _("The harsh sunlight activated\n{B_SCR_ACTIVE_NAME_WITH_PREFIX}'s {B_LAST_ABILITY}!");
 static const u8 sText_StatWasHeightened[] = _("{B_SCR_ACTIVE_NAME_WITH_PREFIX}'s {B_BUFF1} was heightened!");
+static const u8 sText_BoosterEnergyActivates[] = _("{B_SCR_ACTIVE_NAME_WITH_PREFIX}'s Booster Energy\nto activate {B_SCR_ACTIVE_ABILITY}!");
```
Comment (Bassoonian) [inline src/battle_message.c:794]: This is missing a word I think

File: data/battle_scripts_1.s (hunk 1 above)
Comment (AlexOn1ine) [reply data/battle_scripts_1.s:9929]: The script ends with `end2`. Tbh I don't know the difference. Though it seems there is no consistency at all. Some scripts don't even have the suffix even when they should

File: src/battle_message.c (hunk 2 above)
Comment (AlexOn1ine) [reply src/battle_message.c:794]: Not sure which

Comment (Bassoonian) [reply src/battle_message.c:794]: There's no verb in the sentence, which is why I'm confused. Maybe a "caused to activate" or something?

Comment (AlexOn1ine) [reply src/battle_message.c:794]: oh right.

Comment (AlexOn1ine) [reply src/battle_message.c]: fixed

File: src/battle_main.c [hunk 3]
```diff
@@ -3467,6 +3467,7 @@ void SwitchInClearSetData(u32 battler)
     gBattleStruct->lastTakenMoveFrom[battler][3] = 0;
     gBattleStruct->lastMoveFailed &= ~(gBitTable[battler]);
     gBattleStruct->palaceFlags &= ~(gBitTable[battler]);
+    gBattleStruct->boosterEnergyActivates &=  ~(gBitTable[battler]);
```
Comment (Bassoonian) [inline src/battle_main.c:3470]:
```suggestion
    gBattleStruct->boosterEnergyActivates &= ~(gBitTable[battler]);
```

File: src/battle_main.c [hunk 4]
```diff
@@ -3598,6 +3599,7 @@ const u8* FaintClearSetData(u32 battler)
     gBattleStruct->lastTakenMoveFrom[battler][3] = 0;
 
     gBattleStruct->palaceFlags &= ~(gBitTable[battler]);
+    gBattleStruct->boosterEnergyActivates &=  ~(gBitTable[battler]);
```
Comment (Bassoonian) [inline src/battle_main.c:3602]:
```suggestion
    gBattleStruct->boosterEnergyActivates &= ~(gBitTable[battler]);
```

File: test/battle/ability/booster_energy.c [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        else if (spAttack == 110)
+            MESSAGE("RagingBolt's Sp. Atk was heightened!");
+        else if (spDefense == 110)
+            MESSAGE("RagingBolt's Sp. Def was heightened!");
+    } THEN {
+        EXPECT(player->item == ITEM_NONE);
+    }
+}
+
+SINGLE_BATTLE_TEST("Booster Energy activates Quark Drive and increasaes highest stat")
```
Comment (Bassoonian) [inline test/battle/ability/booster_energy.c:101]:
```suggestion
SINGLE_BATTLE_TEST("Booster Energy activates Quark Drive and increases highest stat")
```

File: test/battle/ability/booster_energy.c [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        MESSAGE("RagingBolt's Sp. Atk was heightened!");
+        MESSAGE("The sunlight faded.");
+        ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_HELD_ITEM_EFFECT, player);
+        ABILITY_POPUP(player, ABILITY_PROTOSYNTHESIS);
+        MESSAGE("RagingBolt's used its Booster Energy to activate Protosynthesis!");
+        MESSAGE("RagingBolt's Sp. Atk was heightened!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Booster Energy activates Protosynthesis and increasaes highest stat")
```
Comment (Bassoonian) [inline test/battle/ability/booster_energy.c:67]:
```suggestion
SINGLE_BATTLE_TEST("Booster Energy activates Protosynthesis and increases highest stat")
```

File: test/battle/ability/booster_energy.c [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        TURN { MOVE(opponent, MOVE_ROUND); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_ROUND, opponent);
+        HP_BAR(player, captureDamage: &results[i].damage);
+    } FINALLY {
+        EXPECT_MUL_EQ(results[0].damage, Q_4_12(0.7), results[1].damage);
+    }
+}
+
+SINGLE_BATTLE_TEST("Booster Energy can't be flinged if a Paradox species is involved")
```
Comment (Bassoonian) [inline test/battle/ability/booster_energy.c:185]:
```suggestion
SINGLE_BATTLE_TEST("Booster Energy can't be flung if a Paradox species is involved")
```

File: test/battle/ability/booster_energy.c [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        TURN { MOVE(player, MOVE_ROUND); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_ROUND, player);
+        HP_BAR(opponent, captureDamage: &results[i].damage);
+    } FINALLY {
+        EXPECT_MUL_EQ(results[0].damage, Q_4_12(1.3), results[1].damage);
+    }
+}
+
+SINGLE_BATTLE_TEST("Booster Energy increasaes special defense by 30% if it is the highest stat", s16 damage)
```
Comment (Bassoonian) [inline test/battle/ability/booster_energy.c:159]:
```suggestion
SINGLE_BATTLE_TEST("Booster Energy increases special defense by 30% if it is the highest stat", s16 damage)
```

File: test/battle/ability/booster_energy.c [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+        else if (spAttack == 110)
+            MESSAGE("Iron Moth's Sp. Atk was heightened!");
+        else if (spDefense == 110)
+            MESSAGE("Iron Moth's Sp. Def was heightened!");
+    } THEN {
+        EXPECT(player->item == ITEM_NONE);
+    }
+}
+
+SINGLE_BATTLE_TEST("Booster Energy increasaes special attack by 30% if it is the highest stat", s16 damage)
```
Comment (Bassoonian) [inline test/battle/ability/booster_energy.c:133]:
```suggestion
SINGLE_BATTLE_TEST("Booster Energy increases special attack by 30% if it is the highest stat", s16 damage)
```

**Review by Bassoonian — CHANGES_REQUESTED** [review]

**Review by Bassoonian — APPROVED** [review]

**Pull request merged**